frontend/src/lib/Quiver.ts

```diff
@@ -111,7 +112,110 @@ type IntervalClosed = "left" | "right" | "both" | "neither"
 
 // Our current Typescript version (3.9.5) doesn't support template literal types,
 // so we have to use string literals for now.
-type IntervalIndex = string
+type IntervalType = string
+
+// The frequency strings defined in pandas.
+// See: https://pandas.pydata.org/docs/user_guide/timeseries.html#dateoffset-objects
+type PandasOffsetType =
+  | "B"
+  | "C"
+  | "W"
+  | "WOM"
+  | "LWOM"
+  | "M"
+  | "MS"
+  | "BM"
+  | "BMS"
+  | "CBM"
+  | "CBMS"
+  | "SM"
+  | "SMS"
+  | "Q"
+  | "QS"
+  | "BQ"
+  | "BQS"
+  | "REQ"
+  | "A"
+  | "AS"
+  | "BYS"
+  | "BA"
+  | "BAS"
+  | "RE"
+  | "BH"
+  | "CBH"
+  | "D"
+  | "H"
+  | "T"
+  | "S"
+  | "L"
+  | "U"
+  | "N"
+// type PeriodFrequency = PandasOffsets || `${PandasOffsets}-${string}`
+// type PeriodType = `period[${PeriodFrequency}]`
+
+// Our current Typescript version (3.9.5) doesn't support template literal types,
+// so we have to use string literals for now.
+type PeriodType = string
+type PeriodFrequency = string
+
+const WEEKDAY_SHORT = ["SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT"]
+
+// TODO: For now, we only support the most commonly used offset types.
```

Comment:
Maybe we add this to the tech debt DB to add support for additional offset types.

Reply:
Btw. what happens if unsupported offset types are used?

Reply:
Added this issue to tech debt DB.

In this case, we raise the exception: https://github.com/streamlit/streamlit/blob/d410a0ac8f41ee13a81eda912907e97763869076/frontend/src/lib/Quiver.ts#L839-L841

Reply:
I think doing it this way, it's hard to understand which frequencies are actually supported. How about we:
1. replace `PandasOffsetType` with `SupportedPandasOffsetType`
2. unwrap `Record<SupportedPandasOffsetType, (duration: number, freqParam?: string) => string>` from `Partial`?